runtime/src/iree/io/stream.h

```diff
@@ -0,0 +1,228 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_IO_STREAM_H_
+#define IREE_IO_STREAM_H_
+
+#include <stdint.h>
+
+#include "iree/base/api.h"
+#include "iree/base/internal/atomics.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+//===----------------------------------------------------------------------===//
+// iree_io_stream_t
```

Comment:
Is there a C/C++ standard library type that we could use instead of a custom type? I'm sure there are good reasons for reinventing this wheel, more wondering if you could comment about the (growing) number of wheels that we have in the project. Answer could be as simple as "C++ has this but C doesn't" or "standard versions of this don't allow for custom resource types".

> can either be served from memory (like today), platform file handles, or custom user resources.

So there could be a `FILE` implementation of this interface, for example? (Thinking about how a user of this library code may want to connect their existing data sources to this)

Reply:
not sure I follow - are you referring to iree_io_stream_t being duplicative of something in the STL? there's nothing like this, and even if there was we couldn't use it (not C++), and even if we could use it the design point we have (backing this with mappable memory/external handles that need to manage the lifetime of our ref counted objects/etc) puts us at odds with any such design that would exist there. there's no wheel to reinvent, and if there were we'd be wanting a jet plane and not a horse-drawn carriage ;)

Reply:
Makes sense, thanks. This looked sort of similar to `FILE` from `stdio.h`, yeah.